lambda/value_rec_compiler.ml

```diff
@@ -0,0 +1,778 @@
+      the inner let binding gets translated to code that raises
+      [Match_failure] for non-matching branches.
+      Tracking [Unreachable] explicitly allows us to recover the size
+      of the only non-raising branch. *)
+  | Constant
+  (** Constant values.
+      Can be either an integer-like constant ([0], ['a'], [None],
+      the empty list or the unit constructor), or a structured constant
+      (["hello"], [Some 1], ...).
+
+      Integer constants cannot be pre-allocated, so need their own
+      classification and compilation scheme (See {!Compilation} below).
+      Structured constants could fit into the [Block] category, but we
+      choose to reuse the [constant] classification to avoid sorting
+      through the [Lconst] definitions.
+      It also generates slightly better code. *)
+  | Function
+  (** Function definitions.
+      This includes more than just obvious, syntactic function definitions;
+      see {!Function Lifting} for details. *)
+  | Block of block_size
+  (** Allocated values of a fixed size.
+      This corresponds to expressions ending in a single obvious allocation,
+      but also some more complex expressions where the block is bound to
+      an intermediate variable before being returned.
+  *)
+
+type binding_size = (lambda_with_env, size) Lazy_backtrack.t
+and lambda_with_env = {
+  lambda : lambda;
+  env : binding_size Ident.Map.t;
+}
+
+let dynamic_size () =
+  Misc.fatal_error "letrec: No size found for Static binding"
+
+let join_sizes size1 size2 =
+  match size1, size2 with
+  | Unreachable, size | size, Unreachable -> size
+  | _, _ -> dynamic_size ()
+
+let compute_static_size lam =
+  let rec compute_expression_size env lam =
+    match lam with
+    | Lvar v ->
+      begin match Ident.Map.find_opt v env with
+      | None ->
+        dynamic_size ()
+      | Some binding_size ->
+        Lazy_backtrack.force
+          (fun { lambda; env } -> compute_expression_size env lambda)
+          binding_size
+      end
+    | Lmutvar _ -> dynamic_size ()
+    | Lconst _ -> Constant
+    | Lapply _ -> dynamic_size ()
+    | Lfunction _ -> Function
+    | Llet (_, _, id, def, body) ->
+      let env =
+        Ident.Map.add id (Lazy_backtrack.create { lambda = def; env }) env
+      in
+      compute_expression_size env body
+    | Lmutlet(_, _, _, body) ->
+      compute_expression_size env body
+    | Lletrec (bindings, body) ->
+      let env =
+        List.fold_left (fun env_acc { id; def = _ } ->
+            Ident.Map.add id (Lazy_backtrack.create_forced Function) env_acc)
+          env bindings
+      in
+      compute_expression_size env body
+    | Lprim (p, args, _) ->
+      size_of_primitive env p args
+    | Lswitch (_, sw, _) ->
+      let fail_case =
+        match sw.sw_failaction with
+        | None -> []
+        | Some fail -> [0 (* ignored *), fail]
+      in
+      compute_and_join_sizes_switch env [sw.sw_consts; sw.sw_blocks; fail_case]
```

Comment:
I'm wondering if this code can actually be executed. `Value_rec_check.classify_expression` says that `Texp_match` is `Dynamic` (same with other cases, such as `Texp_ifthenelse`), so wouldn't we expect static bindings to never contain these cases? It might be interesting to add tests to the testsuite that exercise all the cases.

Reply:
My answer to your good question is three-fold:

1. Branching can be introduced between the typedtree and the Lambda code, for example on non-exhaustive let bindings; we were reminded of this in #12302, which discusses possible fixes, and then fixed by @lthls in #12059 
2. As you point out in your comment on `join_sizes`, it would make sense to accept more definitions, for example those that return the same (non-Unreachable) size in all cases, or those that use branches with unreachable cases in the typedtree already (by extending the classification). In general I think that the code in the present PR should be "reasonably general": not be too tightly coupled to the specific set of terms accepted by Rec_check (so that we don't always have to change both in lockstep), but not more complex for the sake of generality.
3. The design goal was never to accept more recursive values, so the PR is not trying to extend Rec_check to benefit from the extra generality in compiler when there is some. I think this is the right choice, but it means that we will indeed end up in some gap between what the type-checker accepts and what the compiler would support, with examples that would work correctly but cannot be written down at the source level.

Reply:
This code was initially written with eager computation of sizes for let bindings, so it could be reached in cases like:
```ocaml
let rec l =
  let v = match foo with
    ...
  in
  v :: l
```
With the lazy environments, this code has very likely become unreachable, but I think it's less work to keep the code as it is than to be sure it is actually unreachable.